File: cpp/examples/cryptexd.cpp
// Jackson Coxson
// Talk to cryptexd, the daemon that installs the DeveloperDiskImage cryptex on
// iOS 17+.

#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include <idevice++/core_device_proxy.hpp>
#include <idevice++/cryptexd.hpp>
#include <idevice++/ffi.hpp>
#include <idevice++/provider.hpp>
#include <idevice++/rsd.hpp>
#include <idevice++/usbmuxd.hpp>

using namespace IdeviceFFI;

[[noreturn]]
static void die(const char* msg, const FfiError& e) {
    std::cerr << msg << ": " << e.message << " (" << e.code << ")\n";
    std::exit(1);
}

static void usage(const char* argv0) {
    std::cerr << "Usage:\n"
              << "  " << argv0 << " list\n"
              << "  " << argv0 << " nonce\n"
              << "  " << argv0 << " ddi\n"
              << "  " << argv0 << " install-ddi <restore_dir>\n"
              << "  " << argv0 << " uninstall <identifier> [version]\n";
}

static void print_hex(const std::vector<uint8_t>& data) {
    for (uint8_t byte : data) {
        std::printf("%02x", byte);
    }
    std::printf("\n");
}

int main(int argc, char** argv) {
    if (argc < 2) {
        usage(argv[0]);
        return 2;
    }
    const std::string cmd     = argv[1];

    auto              mux     = UsbmuxdConnection::default_new(0).expect("failed to reach usbmuxd");
    auto              devices = mux.get_devices().expect("failed to list devices");
    if (devices.empty()) {
        std::cerr << "no devices connected\n";
        return 1;
    }
    auto& dev    = devices[0];
    auto  udid   = dev.get_udid();
    auto  mux_id = dev.get_id();
    if (udid.is_none() || mux_id.is_none()) {
        std::cerr << "device has no UDID or mux id\n";
        return 1;
    }

    auto provider =
        Provider::usbmuxd_new(UsbmuxdAddr::default_new(), 0, udid.unwrap(), mux_id.unwrap(), "cryptexd")
            .expect("failed to create provider");

    auto cdp = CoreDeviceProxy::connect(provider).unwrap_or_else(
        [](FfiError e) -> CoreDeviceProxy { die("failed to connect CoreDeviceProxy", e); });
    auto rsd_port = cdp.get_server_rsd_port().unwrap_or_else(
        [](FfiError e) -> uint16_t { die("failed to get the RSD port", e); });
    auto adapter = std::move(cdp).create_tcp_adapter().expect("failed to create the tunnel");
    auto stream  = adapter.connect(rsd_port).expect("failed to connect the RSD stream");
    auto rsd     = RsdHandshake::from_socket(std::move(stream)).expect("failed the RSD handshake");

    // The daemon serves one routine per connection, so each command below
    // connects its own client.
    if (cmd == "list") {
        auto client = Cryptexd::connect_rsd(adapter, rsd).unwrap_or_else([](FfiError e) -> Cryptexd {
            die("failed to connect cryptexd", e);
        });
        auto installed = client.copy_installed().unwrap_or_else(
            [](FfiError e) -> std::vector<InstalledCryptex> { die("copy_installed failed", e); });
        for (const auto& cryptex : installed) {
            std::cout << cryptex.identifier << " " << cryptex.version << "\n";
        }
        return 0;
    }

    if (cmd == "nonce") {
        auto client = Cryptexd::connect_rsd(adapter, rsd).unwrap_or_else([](FfiError e) -> Cryptexd {
            die("failed to connect cryptexd", e);
        });
        auto nonce = client.get_nonce(NonceDomain::cryptex())
                         .unwrap_or_else([](FfiError e) -> std::vector<uint8_t> {
                             die("get_nonce failed", e);
                         });
        print_hex(nonce);
        return 0;
    }

    if (cmd == "ddi") {
        auto installed = cryptexd_installed_ddi(adapter, rsd)
                             .unwrap_or_else([](FfiError e) -> Option<InstalledCryptex> {
                                 die("installed_ddi failed", e);
                             });
        if (installed.is_none()) {
            std::cout << "no DDI installed\n";
            return 1;
        }
        auto ddi = installed.unwrap();
        std::cout << ddi.identifier << " " << ddi.version << "\n";
        return 0;
    }

    if (cmd == "install-ddi") {
        if (argc < 3) {
            usage(argv[0]);
            return 2;
        }
        auto assets = Cryptex1Assets::load(argv[2]).unwrap_or_else([](FfiError e) -> Cryptex1Assets {
            die("failed to load the DDI assets", e);
        });
        auto installed = cryptexd_install_ddi(adapter, rsd, assets)
                             .unwrap_or_else([](FfiError e) -> InstalledCryptex {
                                 die("failed to install the DDI", e);
                             });
        std::cout << "installed " << installed.identifier << " " << installed.version << "\n";
        return 0;
    }

    if (cmd == "uninstall") {
        if (argc < 3) {
            usage(argv[0]);
            return 2;
        }
        auto client = Cryptexd::connect_rsd(adapter, rsd).unwrap_or_else([](FfiError e) -> Cryptexd {
            die("failed to connect cryptexd", e);
        });
        Option<std::string> version;
        if (argc >= 4) {
            version = std::string(argv[3]);
        }
        auto res = client.uninstall(argv[2], version);
        if (res.is_err()) {
            die("uninstall failed", res.unwrap_err());
        }
        std::cout << "uninstalled " << argv[2] << "\n";
        return 0;
    }

    usage(argv[0]);
    return 2;
}

File: tools/src/cryptexd.rs
// Jackson Coxson

use idevice::{
    IdeviceService, ReadWrite, RsdService,
    core_device_proxy::CoreDeviceProxy,
    cryptexd::{Cryptex1Assets, CryptexdClient, NonceDomain, install_ddi, installed_ddi},
    provider::{IdeviceProvider, RsdProvider},
    rsd::RsdHandshake,
};
use jkcli::{CollectedArguments, JkArgument, JkCommand, JkFlag};
use plist_macro::pretty_print_dictionary;

pub fn register() -> JkCommand {
    let nonce_flags = |cmd: JkCommand| {
        cmd.with_flag(
            JkFlag::new("index")
                .with_help("Nonce domain index (default: 2)")
                .with_argument(JkArgument::new().required(true)),
        )
        .with_flag(
            JkFlag::new("handle")
                .with_help("Nonce domain handle, e.g. a build identity's Cryptex1,NonceDomain")
                .with_argument(JkArgument::new().required(true)),
        )
    };

    JkCommand::new()
        .help("Query and install cryptexes over cryptexd")
        .with_subcommand(
            "list",
            JkCommand::new().help("List the cryptexes installed on the device"),
        )
        .with_subcommand(
            "identifiers",
            JkCommand::new().help("Read the device's AppleImage4 personalization identifiers"),
        )
        .with_subcommand(
            "nonce",
            nonce_flags(JkCommand::new().help("Read a nonce domain's nonce")),
        )
        .with_subcommand(
            "roll-nonce",
            nonce_flags(JkCommand::new().help(
                "Roll a nonce domain's nonce - the domain is unusable until the device reboots",
            )),
        )
        .with_subcommand(
            "uninstall",
            JkCommand::new()
                .help("Uninstall an installed cryptex")
                .with_argument(
                    JkArgument::new()
                        .with_help("cryptex identifier, e.g. com.apple.MobileAsset.DDI")
                        .required(true),
                )
                .with_argument(JkArgument::new().with_help("version to scope the removal to")),
        )
        .with_subcommand(
            "install-ddi",
            JkCommand::new()
                .help("Personalize and install the DeveloperDiskImage cryptex")
                .with_argument(
                    JkArgument::new()
                        .with_help(
                            "unpacked DDI Restore directory, e.g. \
                             /Library/Developer/DeveloperDiskImages/iOS_DDI/Restore",
                        )
                        .required(true),
                )
                .with_flag(
                    JkFlag::new("force")
                        .with_help("Install even when a DDI cryptex is already installed"),
                ),
        )
        .subcommand_required(true)
}

pub async fn main(arguments: &CollectedArguments, provider: Box<dyn IdeviceProvider>) {
    let (sub_name, sub_args) = arguments
        .first_subcommand()
        .expect("no subcommand passed, pass -h for help");
    let mut sub_args = sub_args.clone();

    let proxy = CoreDeviceProxy::connect(&*provider)
        .await
        .expect("no core device proxy");
    let rsd_port = proxy.tunnel_info().server_rsd_port;
    let adapter = proxy.create_software_tunnel().expect("no software tunnel");
    let mut adapter = adapter.to_async_handle();
    let stream = adapter.connect(rsd_port).await.expect("no RSD connect");
    let mut handshake = RsdHandshake::new(stream).await.unwrap();

    if let Some(service) = handshake.services.get("com.apple.security.cryptexd.remote")
        && let Some(features) = &service.features
    {
        eprintln!("cryptexd features: {}", features.join(", "));
    }

    match sub_name.as_str() {
        "list" => {
            let installed = client(&mut adapter, &mut handshake)
                .await
                .copy_installed()
                .await
                .expect("copy-installed failed");
            if installed.is_empty() {
                println!("no cryptexes installed");
            }
            for cryptex in installed {
                println!("{} {}", cryptex.identifier, cryptex.version);
            }
        }
        "identifiers" => {
            let identifiers = client(&mut adapter, &mut handshake)
                .await
                .read_personalization_identifiers()
                .await
                .expect("read-personalization-id failed");
            println!("{}", pretty_print_dictionary(&identifiers));
        }
        "nonce" => {
            let domain = nonce_domain(&mut sub_args);
            let nonce = client(&mut adapter, &mut handshake)
                .await
                .get_nonce(domain)
                .await
                .expect("get-nonce failed");
            println!("structure ({} bytes): {}", nonce.len(), hex(&nonce));
            match idevice::cryptexd::unwrap_nonce(&nonce) {
                Ok(unwrapped) => println!("nonce ({} bytes): {}", unwrapped.len(), hex(&unwrapped)),
                Err(e) => eprintln!("could not unwrap the nonce: {e}"),
            }
        }
        "roll-nonce" => {
            let domain = nonce_domain(&mut sub_args);
            client(&mut adapter, &mut handshake)
                .await
                .roll_nonce(domain)
                .await
                .expect("roll-nonce failed");
            println!("rolled; this domain has no readable nonce until the device reboots");
        }
        "uninstall" => {
            let identifier: String = sub_args.next_argument().expect("identifier required");
            let version = sub_args.next_argument::<String>();
            client(&mut adapter, &mut handshake)
                .await
                .uninstall(&identifier, version.as_deref())
                .await
                .expect("uninstall failed");
            println!("uninstalled {identifier}");
        }
        "install-ddi" => {
            let restore_dir: String = sub_args
                .next_argument()
                .expect("restore directory required");
            let assets = Cryptex1Assets::load(&restore_dir)
                .await
                .expect("failed to load the DDI assets");
            println!(
                "loaded DDI assets: image {} bytes, trustcache {} bytes, info {} bytes, volume hash {} bytes",
                assets.image.len(),
                assets.trustcache.len(),
                assets.info.len(),
                assets.volumehash.len()
            );
            println!(
                "nonce domain handle {}",
                assets.nonce_domain().expect("no nonce domain")
            );

            if !sub_args.has_flag("force")
                && let Some(installed) = installed_ddi(&mut adapter, &mut handshake)
                    .await
                    .expect("copy-installed failed")
            {
                println!(
                    "{} {} is already installed (pass --force to install anyway)",
                    installed.identifier, installed.version
                );
                return;
            }

            let installed = install_ddi(&mut adapter, &mut handshake, &assets)
                .await
                .expect("install failed");
            println!("installed {} {}", installed.identifier, installed.version);
        }
        _ => unreachable!(),
    }
}

/// cryptexd serves one routine per connection, so every call gets a fresh one.
async fn client(
    adapter: &mut impl RsdProvider,
    handshake: &mut RsdHandshake,
) -> CryptexdClient<Box<dyn ReadWrite>> {
    CryptexdClient::connect_rsd(adapter, handshake)
        .await
        .expect("no cryptexd service")
}

fn nonce_domain(args: &mut CollectedArguments) -> NonceDomain {
    if let Some(handle) = args.get_flag::<u64>("handle") {
        NonceDomain::Handle(handle)
    } else if let Some(index) = args.get_flag::<u64>("index") {
        NonceDomain::Index(index)
    } else {
        NonceDomain::default()
    }
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}
